torch/library.py

```diff
@@ -49,9 +49,43 @@ def fallthrough_kernel():
     raise NotImplementedError("fallthrough_kernel() should never be called.")
 
 
-def Schema(name: str, function: Callable, mutates_args=()) -> str:
+def infer_schema(
+    function: Callable, name: Optional[str] = None, mutates_args=()
+) -> str:
+    r"""Parses the schema of a given function with type hints. The schema is inferred from the
+        function's type hints, and can be used to define a new operator.
+
+    We make the following assumptions:
+    - None of the outputs alias any of the inputs or each other.
+    - String type annotations "device, dtype, Tensor, types" without library specification
+      are assumed to be torch.*. Similarly, string type annotations "Optional, List, Sequence, Union"
+      without library specification are assumed to be typing.*.
+    - Only the args listed in ``mutates_args`` are being mutated. If ``mutates_args`` is "unknown",
+        it assumes that all inputs to the operator are being mutates.
+
+    Callers (e.g. the custom ops API) are responsible for checking these assumptions.
+
+    Args:
+        function: function schema to define a new operator.
+        name (Optional[str]): The name of the operator in the schema. If ``name`` is None, then the
+                              name of the function is used as the name of the operator.
```

Comment:
I'd just return the schema without a name, like "(Tensor x) -> Tensor)" instead of "foo(Tensor x) -> Tensor). It's a bit dangerous to unconditionally do `fn.__name__` (this can be <lambda> for lambas and have dots in it for non-global functions)

Reply:
If we return the schema without a name, then the user needs to use the schema like below. They need to add the name explicitly for m.define to work. I think this might be a little confusing for users? Alternatively, we can make `name` required, so we never infer it from function names. 

```
m.define("foo"+torch.library.infer_schema(function=foo_impl, name="foo", mutates_args={}))
```

Or maybe this is fine, and I'll just remove the name.

Reply:
Name required seems fine; does that run us into any problems if we combine torch._library.infer_schema and torch.library.infer_schema?

Reply:
I made a new version, where if op_name is None, we do not include it.

Reply:
> Name required seems fine; does that run us into any problems if we combine torch._library.infer_schema and torch.library.infer_schema?

I made it optional instead of required now, because if we change the old `infer_schema` and make `op_name` required, we'd have to change all the tests and everywhere `infer_schema` was used. if we really want to make `op_name` required, maybe we should do it in a separate PR following this one.

Reply:
It's up to you